### PR TITLE
Executors: enable dequeueing and heartbeating for multiple queues

### PR DESCRIPTION
**Pull request by sanderginn**
- Closes https://github.com/sourcegraph/sourcegraph/issues/50614

Depends on the following issues:
- https://github.com/sourcegraph/sourcegraph/issues/50616
  - Merged in https://github.com/sourcegraph/sourcegraph/pull/52020
- https://github.com/sourcegraph/sourcegraph/issues/51656
  - Merged in https://github.com/sourcegraph/sourcegraph/pull/52381
- https://github.com/sourcegraph/sourcegraph/issues/51658
  - Merged in  https://github.com/sourcegraph/sourcegraph/pull/52525

Due to the dependency on the work listed above, all PRs were ultimately merged into this branch, so the client side implementation got a little bit messy to review, for which I apologize. The initial intent of this PR, closing #50614, was reviewed before the merges occurred.

#### batches, single queue
https://github.com/sourcegraph/sourcegraph/assets/2979513/6d325326-d4f5-4e6f-a5e0-0918dbde07bf

#### codeintel, single queue
https://github.com/sourcegraph/sourcegraph/assets/2979513/ff3a87db-ef07-4221-a159-73b93a6ae55b

#### batches, single queue
https://github.com/sourcegraph/sourcegraph/assets/2979513/c66a9b9f-b745-4abc-bf87-9077dd5f2959

#### codeintel, single queue
https://github.com/sourcegraph/sourcegraph/assets/2979513/785d7f34-630f-4c2a-abb6-82ca3b896813

#### batches, single queue 
https://github.com/sourcegraph/sourcegraph/assets/2979513/69c5c06b-ea1c-47f7-9196-e3f108794493

#### codeintel, single queue
https://github.com/sourcegraph/sourcegraph/assets/2979513/4e791d98-06cd-4fa7-93dd-81be0a73c744

#### batches + codeintel in parallel, multi queue
https://github.com/sourcegraph/sourcegraph/assets/2979513/004608dd-f98c-4305-aeb3-a01444c12224

## Initial PR description (enable client to dequeue from multiple queues)
This PR updates the client's `Dequeue` method to dequeue from multiple queues. It's backwards compatible with single-queue configurations. The field `Job.Queue` is only set when dequeuing from the general `/dequeue` endpoint, so the job `MarkXxx` methods default to single-queue behaviour if the job does not specify a queue name.

Unrelated to the linked issue, the multi-handler will log and return no content early in the event of an empty queue list in a dequeue request (although this should never occur).

## Test plan
- [x] Unit tests
- [x] Local testing
- [x] Dogfood testing
- [x] Demo

### CODE REVIEW

File: enterprise/cmd/frontend/internal/executorqueue/handler/multihandler.go [hunk 1]
```diff
@@ -85,6 +85,11 @@ func (m *MultiHandler) dequeue(ctx context.Context, req executortypes.DequeueReq
 		}
 	}
 
+	if len(req.Queues) == 0 {
```
Comment (Piszmog) [inline enterprise/cmd/frontend/internal/executorqueue/handler/multihandler.go:88]: This already covered by a unit test, right?

**Review by Piszmog — APPROVED** [review]
Looks good! Feels too straight forward 👀 

Could you also run a main dry run? I would like to ensure the E2E tests pass

File: enterprise/cmd/frontend/internal/executorqueue/handler/multihandler.go (hunk 1 above)
Comment (sanderginn) [reply enterprise/cmd/frontend/internal/executorqueue/handler/multihandler.go:88]: Now it is 😄

File: enterprise/cmd/executor/internal/apiclient/queue/client.go [hunk 2]
```diff
@@ -150,6 +167,12 @@ func (c *Client) MarkFailed(ctx context.Context, job types.Job, failureMessage s
 }
 
 func (c *Client) Heartbeat(ctx context.Context, jobIDs []int) (knownIDs, cancelIDs []int, err error) {
+	if len(c.options.QueueNames) > 0 {
+		// TODO: multi-queue heartbeats are not implemented yet, so simply return the job ids immediately
+		// to allow jobs to terminate while testing
+		return jobIDs, nil, nil
+	}
```
Comment (sanderginn) [inline enterprise/cmd/executor/internal/apiclient/queue/client.go:174]: @Piszmog this was the best workaround I could come up with, but still breaks when processing jobs from a queue that isn't defined first in the list.

I propose to not merge this PR until the heartbeat is implemented.

Comment (Piszmog) [reply enterprise/cmd/executor/internal/apiclient/queue/client.go:174]: Sounds good. Let's hold off on this PR until you have the heartbeat merged.

**Comment by sourcegraph-buildkite** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.00% (+0.46 kb) | 0.00% (+0.46 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/sginn/executor-multi-dequeue-client/compare-report.html;tab=live_object) for a full comparison between the commits e18cc37 and 42be886 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Review by Piszmog — APPROVED** [review]
Looks great! This is awesome!

Just one minor question on terraform.

File: doc/admin/executors/deploy_executors_terraform.md [hunk 3]
```diff
@@ -33,22 +33,28 @@ module "executors" {
 
   executor_sourcegraph_external_url            = "<external url>"
   executor_sourcegraph_executor_proxy_password = "<shared secret>"
+              
+  # Either:
   executor_queue_name                          = "<codeintel | batches>"
+  # Or:
+  executor_queue_names                         = "codeintel,batches"
```
[the remark this answers is not in the document]
Comment (sanderginn) [reply doc/admin/executors/deploy_executors_terraform.md:40]: Fair point 👍

**Comment by sanderginn** [comment]
Completed testing, added demos.

All that is left is a changelog entry!

**Pull request merged**